src/js/select2/core.js

```diff
@@ -94,7 +94,8 @@ define([
     if ($element.attr('id') != null) {
       id = $element.attr('id');
     } else if ($element.attr('name') != null) {
-      id = $element.attr('name') + '-' + Utils.generateChars(2);
+      id = $element.attr('name').replace(/\[|\]/g, '');
```

Comment:
jQuery [recommends](https://learn.jquery.com/using-jquery-core/faq/how-do-i-select-an-element-by-an-id-that-has-characters-used-in-css-notation/) using the following regex

```js
/(:|\.|\[|\]|,)/g
```

As it will handle a couple of other characters that can also be valid selectors.